### PR TITLE
[fix] correct logic k-NN algos ```kneighbors()``` call when ```algorithm='brute'``` and fit with GPU

### PR DESCRIPTION
**Pull request by icfaust**
### Description

Calling k-NN algorithms ```kneighbors()``` after fitting with GPU and algorithm = 'brute' will assume that it was fit using daal4py, and will fail in yielding kneighbors.  This adds corrections in that check, and a test for ```kneighbors``` without input.

```kneighbors``` is a bit special because it doesn't require an input to yield numerical results (re-using the fitted `X` values).  The fit can occur in the daal4py backend or onedal backend.  The check for ```predict``` and ```kneighbors``` now will check for which train object was generated, and will use it.

---

Checklist to comply with before moving PR from draft:

**PR completeness and readability**

- [x] I have reviewed my changes thoroughly before submitting this pull request.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have updated the documentation to reflect the changes or created a separate PR with update and provided its number in the description, if necessary.
- [x] Git commit message contains an appropriate signed-off-by string _(see [CONTRIBUTING.md](https://github.com/intel/scikit-learn-intelex/blob/main/CONTRIBUTING.md#pull-requests) for details)_.
- [x] I have added a respective label(s) to PR if I have a permission for that.  
- [x] I have resolved any merge conflicts that might occur with the base branch.

**Testing**

- [x] The unit tests pass successfully.
- [x] I have run it locally and tested the changes extensively.

**Performance**

- [x] I have measured performance for affected algorithms using [scikit-learn_bench](https://github.com/IntelPython/scikit-learn_bench) and provided at least summary table with measured data, if performance change is expected.
- [x] I have provided justification why performance has changed or why changes are not expected.

### CODE REVIEW

**Comment by icfaust** [comment]
onedal knn implementation is a nightmare

File: onedal/neighbors/neighbors.py [hunk 1]
```diff
@@ -597,21 +595,19 @@ def _onedal_fit(self, X, y, queue):
...
 
             return predict_alg(**params).compute(X, model)
 
+        gpu_device = queue is not None and queue.sycl_device.is_gpu
         policy = self._get_policy(queue, X)
         X = _convert_to_supported(policy, X)
         backend = (
             self._get_backend("neighbors", "regression", None)
-            if gpu_device
+            if queue is not None and queue.sycl_device.is_gpu
```
Comment (samir-nasibli) [inline onedal/neighbors/neighbors.py:610]: please reuse `gpu_device` here.
```suggestion
            if gpu_device
```

**Review by samir-nasibli — COMMENTED** [review]
Nice work @icfaust.
Just a questions: could we validate the perf, since the flow is changed a little bit?

**Review by samir-nasibli — APPROVED** [review]
@icfaust I am good with the changes proposed. Just minor comments
I recommend wait CI back for the last validation

File: sklearnex/neighbors/tests/test_neighbors.py [hunk 2]
```diff
@@ -52,20 +52,22 @@ def test_sklearnex_import_knn_regression(dataframe, queue):
...
     X = [[0, 0, 2], [1, 0, 0], [0, 0, 1]]
     X = _convert_to_dataframe(X, sycl_queue=queue, target_df=dataframe)
     test = _convert_to_dataframe([[0, 0, 1.3]], sycl_queue=queue, target_df=dataframe)
-    neigh = estimator(n_neighbors=2).fit(X)
+    neigh = estimator(n_neighbors=2, algorithm=algorithm).fit(X)
     result = neigh.kneighbors(test, 2, return_distance=False)
     result = _as_numpy(result)
     assert "sklearnex" in neigh.__module__
     assert_allclose(result, [[2, 0]])
+    result = neigh.kneighbors()
```
[the remark this answers is not in the document]
Comment (icfaust) [reply sklearnex/neighbors/tests/test_neighbors.py:70]: All use of ```kneighbors``` in ```test_neighbors``` passes a value to kwarg ```X```.  The datatype and queue are always extracted from it.  However, that doesn't cover the case when the default X=None is used.  This addition makes sure that the default use of ```kneighbors``` runs, though the output is not checked via an ```assert_allclose```.  This closes a testing gap.

[the remark this answers is not in the document]
Comment (samir-nasibli) [reply sklearnex/neighbors/tests/test_neighbors.py:70]: @icfaust I think make sense to add notes in comment section, since no that obvious in the code base.

[the remark this answers is not in the document]
Comment (icfaust) [reply sklearnex/neighbors/tests/test_neighbors.py:70]: Will do!

**Review by Alexsandruss — APPROVED** [review]
